hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocationServiceSupport.java

```diff
@@ -176,6 +178,9 @@ public void cleanResources(ConstructorFunction<Object, Throwable> responseCtor,
                 invocation.notifyException(responseCtor.createNew(null));
             }
         }
+        //Try cleaning again later to compensate the Race between `send` and `cleanResources`
```

Comment:
Be careful with using a shared thread because you don't know who is using this pool and it could be depleted for some time and then user invocations don't get triggered in case of a closed connection.

I would introduce a dedicated thread that monitors the invocations; the same as the serverside invocation monitor.